Add last refresh status for provider dashboards

kavyanekkalapu opened the pull request:
Issue: https://github.com/ManageIQ/manageiq-ui-classic/issues/7709

After:
<img width="1502" alt="Screen Shot 2021-04-23 at 2 13 00 PM" src="https://user-images.githubusercontent.com/37085529/115912999-08f2f580-a43e-11eb-83df-34c9f370f361.png">

@miq-bot assign @Fryguy
@miq-bot add-reviewer @Fryguy @agrare

Fryguy commented on app/helpers/textual_mixins/textual_refresh_status.rb:
```diff
@@ -1,20 +1,29 @@
+  end
+
+  def refresh_status(record)
+    last_refresh_status = record.last_refresh_status ? record.last_refresh_status.titleize : nil
```
Simplified, this can be

```suggestion
    last_refresh_status = record.last_refresh_status&.titleize
```

Fryguy commented on app/helpers/textual_mixins/textual_refresh_status.rb:
```diff
@@ -1,20 +1,29 @@
-  def textual_refresh_date
-    last_refresh_date = @record.last_refresh_date ? @record.last_refresh_date.in_time_zone(Time.zone) : nil
+  def refresh_date(record)
+    p "record.last_refresh_date-------> #{record.name}"
```
Heads up on debug code.

kavyanekkalapu replied:
@Fryguy this is WIP PR :)

Fryguy commented on app/helpers/textual_mixins/textual_refresh_status.rb:
```diff
@@ -1,20 +1,29 @@
-    last_refresh_date = @record.last_refresh_date ? @record.last_refresh_date.in_time_zone(Time.zone) : nil
+  def refresh_date(record)
+    p "record.last_refresh_date-------> #{record.name}"
+    last_refresh_date = record.last_refresh_date ? record.last_refresh_date.in_time_zone(Time.zone) : nil
```
Can also be written as:

```suggestion
   last_refresh_date = record.last_refresh_date && record.last_refresh_date.in_time_zone(Time.zone)
```

chessbyte replied:
@Fryguy can't it just be `last_refresh_date = record.last_refresh_date&.in_time_zone(Time.zone)` ?

Fryguy commented on app/javascript/components/refersh-notifications.jsx:
```diff
@@ -0,0 +1,53 @@
+import React from 'react';
```
typo in the file name?  `refersh`

Fryguy commented on app/services/container_dashboard_service.rb:
```diff
@@ -49,6 +51,14 @@ def image_metrics_data
+    {
+      :defaultCred     => ( @ems.present? && @ems.enabled? ) ? 'Valid': 'Invalid',
+      :lastRefreshDate => @ems.present? ? refresh_date(@ems) : [],
+      :lastRefresh     => @ems.present? ? refresh_status(@ems) : []
```
Ruby is usually in snake_case, so this should be `last_refresh` (and similar changes elsewhere in the ruby code (unless there was an interop reason with JS)

kavyanekkalapu replied:
@Fryguy means, lastRefreshDate is last_refersh_date?, it seems to be working fine.

Fryguy commented:
@kavyanekkalapu This is coming along nice!  Do you have screenshots (or still too early yet?)

kavyanekkalapu commented:
@Fryguy This is WIP PR, i will assign it to you once it is ready :)

kavyanekkalapu commented:
@Fryguy @agrare so far i foxed cloud, infra, physical infra and container providers, am i i missing any other providers here?

chessbyte commented:
from the screenshots, if we can show Last Refresh Status with a relative time (22 days ago or almost 3 years ago), do we even need the Last Refresh Date?

kavyanekkalapu commented:
@Fryguy @agrare This pr is ready for review, just listing out the scenario in which notification will show up

- Default credentials status is anything other than `valid`
- Last_refresh_status is anything other than `success`
- Last refresh date is available
- Last refresh date is with in last 2 days.
If any provider pass all above four  conditions, we will show up notification.

Please review whenever you have time, thanks.

Fryguy commented on app/helpers/textual_mixins/textual_refresh_status.rb:
```diff
@@ -1,20 +1,32 @@
+                      {:value => record.last_refresh_error.try(:truncate, 120)}],
+      :title      => record.last_refresh_error,
+      :status     => record.last_refresh_status&.titleize,
+      :date_range => !record.last_refresh_date ? false : refresh_time >= before_time # checking if date is within last two days range.
```
Rails 6 has added .before? and .after? methods, so I think using that will make this easier to read.  Additionally, let's avoid the double negatives to make this easier to read, as well as use the Rails date helpers.

```suggestion
      :date_range => refresh_time&.before?(2.days.ago)
```

(Note: if nil is allowed here you can use the following, otherwise you'll have to do `refresh_time && refresh_time.before?...`)

Then above,

```diff
-    before_time = Date.today - 2
-    refresh_time = record.last_refresh_date ? record.last_refresh_date.in_time_zone(Time.zone).strftime("%Y-%m-%d").to_date : nil
-    if record.last_refresh_date
-      last_refresh_date = !record.last_refresh_date ? _('Never') : time_ago_in_words(record.last_refresh_date.in_time_zone(Time.zone)).titleize
+    refresh_time = record.last_refresh_date&.in_time_zone(Time.zone)
+    if refresh_time
+      last_refresh_date = refresh_time ? refresh_time.titleize : _('Never')
```

Fryguy commented on app/helpers/textual_mixins/textual_refresh_status.rb:
```diff
@@ -1,20 +1,32 @@
+      :value      => [{:value => last_refresh_status},
+                      {:value => record.last_refresh_error.try(:truncate, 120)}],
+      :title      => record.last_refresh_error,
+      :status     => record.last_refresh_status&.titleize,
```
Not sure what this is used for but is this essentially a duplicate of the value (but without the date attached?)

Fryguy commented on app/services/ems_cloud_dashboard_service.rb:
```diff
@@ -15,9 +15,10 @@ def aggregate_status_data
-      :attrData => attributes_data,
+      :quadicon      => @controller.instance_exec(@ems, &EmsInfraDashboardService.quadicon_calc),
+      :status        => status_data,
+      :refreshStatus => refresh_data,
```
```suggestion
      :refresh_status => refresh_data,
```

Again not sure on javascript vs Ruby conventions here.

kavyanekkalapu replied:
@fryguy it is working, in fact several other places are using camelcase in ruby files to get data in react( Wg: https://github.com/ManageIQ/manageiq-ui-classic/blob/634700500a5aa8586f08dc6470dddb23fb551ca6/app/services/ems_infra_dashboard_service.rb#L34) , let me know if this is not recommended, will change it.

Fryguy replied:
It's fine, just confusing for me since I'm not used to the react side of things...as long as we are consistent I'm fine with it.

Fryguy commented on app/services/container_dashboard_service.rb:
```diff
@@ -1,6 +1,8 @@
 class ContainerDashboardService < DashboardService
   include ContainerServiceMixin
   include Mixins::CheckedIdMixin
+  include ActionView::Helpers::DateHelper
```
This should not be necessary if `TextualMixins::TextualRefreshStatus` itself includes `ActionView::Helpers::DateHelper`

Fryguy commented on app/services/container_dashboard_service.rb:
```diff
@@ -49,6 +51,13 @@ def image_metrics_data
 
+  def refresh_status_data
+    {
+      :default_cred => @ems.present? && @ems.enabled ? 'Valid' : 'Invalid',
```
Can you add parenthesis here to show the intended order of operations?  I know what the answer is, but it could be readable as either of the following and it's confusing otherwise:

```
(A && B) ? X : Y
A && (B ? X : Y)
```

kavyanekkalapu replied:
having parentheses giving a linter error.

== app/services/container_dashboard_service.rb ==
C: 55: 24: Style/TernaryParentheses: Omit parentheses for ternary conditions.

Fryguy replied:
ok, I disagree with rubocop here, but let's keep it removed.

agrare replied:
Yeah if there are multiple boolean conditions I think parens are a good idea, I don't want to have to think about operator precedence when a simple `(A && B) ? X : Y` makes it obvious what is intended.

Fryguy commented on app/services/ems_dashboard_service.rb:
```diff
@@ -1,6 +1,9 @@
   include Mixins::CheckedIdMixin
+  include TextualSummaryHelper
+  include ActionView::Helpers::DateHelper
+  include TextualMixins::TextualRefreshStatus
```
Also for consistency, let's put all of the includes in the same order across the various dashboard files.

kavyanekkalapu replied:
i am not really what that mean, but rearranged the order, let me know if i need t make any more changes.

Fryguy replied:
Oh I was saying just alphabetize unless the order matters...when the order matters, we should just be consistent across all of the services.

Fryguy commented on config/routes.rb:
```diff
@@ -1011,6 +1011,7 @@
         pod_metrics_data
         project_data
         show
+        refresh_status_data
```
I don't think this should be here, as we don't want to expose refresh_status_data as a route.  I think it should just be a private method on the controller.

kavyanekkalapu replied:
@Fryguy here last_refersh_status is end point, that is why added in routes file.

Fryguy replied:
:+1:

Fryguy replied:
Can you move this up alphabetically?

kavyanekkalapu replied:
done.

Fryguy commented on app/helpers/textual_mixins/textual_refresh_status.rb:
```diff
@@ -1,20 +1,32 @@
+
+  def refresh_status(record)
+    last_refresh_status = record.last_refresh_status&.titleize
+    refresh_time = record.last_refresh_date&.in_time_zone(Time.zone)
```
looking at this, I'm not sure it's right, or that it needs `Time.zone`.  We aren't going to present the time in any particular time zone (if we did, I'd say we should use the user's time zone), but instead we are using relative time, which doesn't need a time zone.  I think we should remove it.

kavyanekkalapu replied:
Ok let me try removing that and see

kavyanekkalapu commented on app/helpers/textual_mixins/textual_refresh_status.rb:
```diff
@@ -1,20 +1,32 @@
+                  {:value => record.last_refresh_error.try(:truncate, 120)}],
+      :title  => record.last_refresh_error,
+      :status => record.last_refresh_status&.titleize,
+      :stale  => refresh_time ? refresh_time.before?(2.days.ago) : true # checking if date is within last two days range.
```
@Fryguy removed Time.zone on date, in line no 24, i had to use ternary operator, otherwise if data is null , stale value  is null too.

Fryguy replied:
I was suggesting remove the enture `&.in_time_zone(Time.zone)`

Fryguy replied:
Ah ok, yeah, makes sense.

Fryguy commented on app/services/container_dashboard_service.rb:
```diff
@@ -49,6 +50,13 @@ def image_metrics_data
 
+  def refresh_status_data
+    {
+      :default_cred => @ems.present? && @ems.enabled ? 'Valid' : 'Invalid',
```
```suggestion
      :default_cred => @ems.present? && @ems.enabled ? _('Valid') : _('Invalid'),
```

kavyanekkalapu replied:
@Fryguy we are not showing this text in ui, i need this as flag for condition to show the status or not.

Fryguy replied:
ohhh...interesting - maybe we should just make it a regular boolean then (and maybe change the key to `:default_cred_valid`, otherwise it's confusing.

Fryguy replied:
Actually, I'm also seeing `textual_authentications` used elsewhere, which I think _will_ be translated, so that makes me concerned this won't work on the other screens?

kavyanekkalapu replied:
@Fryguy All other provider dashboards like physical infra, cloud ... etc, has this field as  
`:default_cred => textual_authentications(@resource.first.authentication_for_summary)`- this gives result Valid/Invalid.I have written code in react for this( to show notification for !Valid cases), i want to use same code for container dashboard too, that is why i wrote a line like this .
:default_cred => @ems.present? && @ems.enabled ? 'Valid' : 'Invalid'.

Fryguy replied:
Is that Valid/Invalid from `textual_authentications` translated?

kavyanekkalapu replied:
@Fryguy im not sure, i don't see it in the function, so assuming it is coming from api.
 ```
def textual_authentications(authentications)
    return [{:label => _("Default Authentication"), :title => t = _("None"), :value => t}] if authentications.blank?

    authentications.collect do |auth|
      label = case auth[:authtype]
              when "default"           then _("Default")
              when "metrics"           then _("C & U Database")
              when "amqp"              then _("AMQP")
              when "console"           then _("VMRC Console")
              when "ipmi"              then _("IPMI")
              when "remote"            then _("Remote Login")
              when "smartstate_docker" then _("SmartState Docker")
              when "ws"                then _("Web Services")
              when "ssh_keypair"       then _("SSH Key Pair")
              else;                    _("<Unknown>")
              end

      {:label => _("%{label} Credentials") % {:label => label},
       :value => auth[:status] || _("None"),
       :title => auth[:status_details]}
    end
  end

Fryguy commented on app/services/ems_physical_infra_dashboard_service.rb:
```diff
@@ -1,5 +1,8 @@
 class EmsPhysicalInfraDashboardService < DashboardService
   include Mixins::CheckedIdMixin
+  include ActionView::Helpers::DateHelper
```
if the `ActionView::Helpers::DateHelper` is included in the `TextualMixins::TextualRefreshStatus`, then you shouldn't need it here (unless of course something in this service directly requires the DateHelper)

kavyanekkalapu replied:
Removed it , thanks.

Fryguy reviewed (approved):
Overall, LGTM - I have a couple more mostly code cleanliness comments, but I think this is ready to go.  Would also like @agrare to review again before merge.

Fryguy commented on app/helpers/textual_mixins/textual_refresh_status.rb:
```diff
@@ -1,20 +1,32 @@
+    last_refresh_status = record.last_refresh_status&.titleize
+    refresh_time = record.last_refresh_date
+    if refresh_time
+      last_refresh_date = refresh_time ? time_ago_in_words(refresh_time).titleize : _(Never)
```
```suggestion
      last_refresh_date = refresh_time ? time_ago_in_words(refresh_time).titleize : _("Never")
```

This feels like it should have blown up in test.

kavyanekkalapu replied:
done, fixed it.

agrare commented:
Maybe I'm doing something wrong but I'm not seeing these error notifications:I have a last refresh error:
![Screenshot from 2021-05-04 11-08-47](https://user-images.githubusercontent.com/12851112/117026102-79650680-acc9-11eb-9971-571375bbc102.png)

But not notifications on the main dashboard or details page:
![Screenshot from 2021-05-04 11-08-37](https://user-images.githubusercontent.com/12851112/117026218-98fc2f00-acc9-11eb-9da8-aad0373d7d06.png)

kavyanekkalapu commented:
> Maybe I'm doing something wrong but I'm not seeing these error notifications:I have a last refresh error:
> ![Screenshot from 2021-05-04 11-08-47](https://user-images.githubusercontent.com/12851112/117026102-79650680-acc9-11eb-9971-571375bbc102.png)
> 
> But not notifications on the main dashboard or details page:
> ![Screenshot from 2021-05-04 11-08-37](https://user-images.githubusercontent.com/12851112/117026218-98fc2f00-acc9-11eb-9da8-aad0373d7d06.png)

@agrare your default cred status is Valid, that is why you are not seeing notifications.

kavyanekkalapu commented:
@agrare please check my comment https://github.com/ManageIQ/manageiq-ui-classic/pull/7719#issuecomment-825840815

agrare commented:
Right I thought if the refresh status was anything other than success I should see a notification though?

kavyanekkalapu commented:
> Right I thought if the refresh status was anything other than success I should see a notification though?

in my comment all conditions are and not or, let me know if you want me to remove the default cred status check, i can remove it, thanks.

agrare commented:
Oh, `all conditions are and not or` this is what I was missing.

If the credentials are valid but the refresh is failing we absolutely should see a notification.

kavyanekkalapu commented:
@agrare alright, i will change it and let you know.

kavyanekkalapu commented:
@agrare done, removed default credentials check, please review again.

agrare commented:
I still don't see any notifications on a cloud provider with invalid default credentials or an infra provider with a failed last refresh status...

kavyanekkalapu commented:
@agrare please pull latest and check now

Fryguy commented on app/javascript/components/aggregate_status_card.jsx:
```diff
@@ -49,7 +49,10 @@ const AggregateStatusCard = ({ providerId, providerType }) => {
+      {(data.refreshStatus && data.refreshStatus.default_cred
+       && (data.refreshStatus.default_cred[0].value !== 'Valid'
+        || !data.refreshStatus.last_refresh.stale
+         || data.refreshStatus.last_refresh.status !== 'Success'))
```
it feels like we should move this down in the refreshStatus itself.  I'm thinking something like `data.refreshStatus.overallStatus`, then here you can just look at that one property.

Fryguy replied:
(we can also do this in a follow-up if the current code works - I just don't like seeing too much logic in the UI)

kavyanekkalapu replied:
Agree, we can improve code, i can do a followup pr later.

Fryguy commented on app/javascript/components/refresh-data-notification.jsx:
```diff
@@ -19,7 +19,7 @@ const RefreshDataNotification = ({
   }, []);
 
-  if (data.loading) return null;
+  if (data.loading || (data.status.default_cred === 'Valid' && data.status.last_refresh.stale && data.status.last_refresh.status === 'Success' )) return null;
```
and then also here you can just look that the one property as well.

agrare commented:
@kavyanekkalapu in the case where the last refresh was successful but now the credentials are invalid I think the error we're showing is a little misleading:
![Screenshot from 2021-05-06 14-22-00](https://user-images.githubusercontent.com/12851112/117347036-8bd07300-ae76-11eb-919a-777310350e18.png)

I think this should say instead that the Default credentials are invalid:
![Screenshot from 2021-05-06 14-23-29](https://user-images.githubusercontent.com/12851112/117347136-b02c4f80-ae76-11eb-86cd-6e111371e8ae.png)

kavyanekkalapu commented:
> @kavyanekkalapu in the case where the last refresh was successful but now the credentials are invalid I think the error we're showing is a little misleading:
> ![Screenshot from 2021-05-06 14-22-00](https://user-images.githubusercontent.com/12851112/117347036-8bd07300-ae76-11eb-919a-777310350e18.png)
> 
> I think this should say instead that the Default credentials are invalid:
> ![Screenshot from 2021-05-06 14-23-29](https://user-images.githubusercontent.com/12851112/117347136-b02c4f80-ae76-11eb-86cd-6e111371e8ae.png)



> @kavyanekkalapu in the case where the last refresh was successful but now the credentials are invalid I think the error we're showing is a little misleading:
> ![Screenshot from 2021-05-06 14-22-00](https://user-images.githubusercontent.com/12851112/117347036-8bd07300-ae76-11eb-919a-777310350e18.png)
> 
> I think this should say instead that the Default credentials are invalid:
> ![Screenshot from 2021-05-06 14-23-29](https://user-images.githubusercontent.com/12851112/117347136-b02c4f80-ae76-11eb-86cd-6e111371e8ae.png)

@agrare yes it is very confusing, can you list out the condition you want to notification to display and please mention which ones are or which one is and, thank you.

agrare commented:
Okay so as I understood it we want to show an error message for three different conditions:

1. Default credentials are invalid
OR
2. Last Refresh Status is Error
OR
3. Last Refresh was more than 2 days ago.

The error shown should make sense relative to which issue the provider has.

kavyanekkalapu commented:
@agrare  and I agreed in  `last refresh error || last refresh date is more than 2 days`, will do changes based on that.

kavyanekkalapu commented:
@agrare @Fryguy can any of you merge the pr please? Thanks.

agrare commented:
@kavyanekkalapu this didn't have any dependent PRs right?  Could we backport this to morphy?

kavyanekkalapu commented:
@agrare I fixed a bug in https://github.com/ManageIQ/manageiq-ui-classic/pull/7744, added morphy for both.

agrare commented:
Awesome thanks @kavyanekkalapu !

Fryguy commented:
@kavyanekkalapu A conflict occurred during the backport of this pull request to `morphy`.

If this pull request is based on another pull request that has not been marked for backport, add the appropriate labels to the other pull request. Otherwise, please create a new pull request direct to the `morphy` branch in order to resolve this.

Conflict details:

```diff
diff --cc app/javascript/packs/component-definitions-common.js
index d2d76a6830,475243331c..0000000000
--- a/app/javascript/packs/component-definitions-common.js
+++ b/app/javascript/packs/component-definitions-common.js
@@@ -31,8 -31,14 +31,15 @@@ import NotificationDrawer from '../comp
  import OptimizationList from '../optimization/optimization_list';
  import OpsTenantForm from '../components/ops-tenant-form/ops-tenant-form';
  import OrcherstrationTemplateForm from '../components/orchestration-template/orcherstration-template-form';
 -import PxeImageForm from '../components/pxe-image-type-form';
 -import PxeCustomizationTemplateForm from '../components/pxe-customization-template-form';
  import PxeServersForm from '../components/pxe-servers-form/pxe-server-form';
  import PerformanceChartWidget from '../components/performance-charts';
++<<<<<<< HEAD
++=======
+ import PodsTrendChart from '../components/provider-dashboard-charts/pod-trend-charts';
+ import RecentHostGraph from '../components/provider-dashboard-charts/recent-host-chart';
+ import RecentVmGraph from '../components/provider-dashboard-charts/recent-vm-chart';
+ import RefreshDataNotification from '../components/refresh-data-notification';
++>>>>>>> faac0462d5... Merge pull request #7719 from kavyanekkalapu/add-status-provider-dashboard
  import RegionForm from '../components/region-form';
  import RemoveCatalogItemModal from '../components/remove-catalog-item-modal';
  import RemoveGenericItemModal from '../components/remove-generic-item-modal';
@@@ -90,9 -99,15 +97,15 @@@ ManageIQ.component.addReact('Notificati
  ManageIQ.component.addReact('OpsTenantForm', OpsTenantForm);
  ManageIQ.component.addReact('OptimizationList', OptimizationList);
  ManageIQ.component.addReact('OrcherstrationTemplateForm', OrcherstrationTemplateForm);
 -ManageIQ.component.addReact('PxeImageForm', PxeImageForm);
 -ManageIQ.component.addReact('PxeCustomizationTemplateForm', PxeCustomizationTemplateForm);
  ManageIQ.component.addReact('PxeServersForm', PxeServersForm);
  ManageIQ.component.addReact('PerformanceChartWidget', PerformanceChartWidget);
 -ManageIQ.component.addReact('PodsTrendChart', PodsTrendChart);
  ManageIQ.component.addReact('Quadicon', Quadicon);
++<<<<<<< HEAD
++=======
+ ManageIQ.component.addReact('RecentHostGraph', RecentHostGraph);
+ ManageIQ.component.addReact('RecentVmGraph', RecentVmGraph);
+ ManageIQ.component.addReact('RefreshDataNotification', RefreshDataNotification);
++>>>>>>> faac0462d5... Merge pull request #7719 from kavyanekkalapu/add-status-provider-dashboard
  ManageIQ.component.addReact('RegionForm', RegionForm);
  ManageIQ.component.addReact('RemoveCatalogItemModal', RemoveCatalogItemModal);
  ManageIQ.component.addReact('RemoveGenericItemModal', RemoveGenericItemModal);
diff --cc app/stylesheet/notifications.scss
index dab2315c0f,9200b6d744..0000000000
--- a/app/stylesheet/notifications.scss
+++ b/app/stylesheet/notifications.scss
@@@ -78,12 -36,197 +78,27 @@@
  }
  
  .backgrounded-icon {
 -  display: inline-block;
 -  height: 20px;
 -  width: 20px;
 -  text-align: center;
 -  i {
 -    align-self: center;
 -    color: #fff;
 -  }
 -}
 -
 -.notification-drawer {
 -  background-color: #fafafa;
 -  border: 1px solid #d1d1d1;
 -  -webkit-box-shadow: 0 6px 12px rgb(3 3 3 / 18%);
 -  box-shadow: 0 6px 12px rgb(3 3 3 / 18%);
 -  overflow-y: auto;
 -  position: absolute;
 -  right: 0;
 -  width: 450px;
 -  z-index: 2;
 -  display: flex;
 -  flex-direction: column;
 -  overflow-y: hidden;
 -  top: 38px;
 -  height: calc(100% - 44px);
 -  .panel-header {
 -    background-color: #fafafa;
 -    border-bottom: 1px solid #d1d1d1;
 -    position: absolute;
 -    width: 100%; 
 -    background-image: linear-gradient(to bottom, #fafafa 0%, #ededed 100%);
 -    .closeIcon {
 -      right: 0px;
 -    }
 -    .collapsed {
 -      left: 0px;
 -      padding-left: 5px;
 -    }
 -    .expanded {
 -      left: 0px;
 -      padding-left: 5px;
 -    }
 -    .heading-icon {
 -      color: #252525;
 -      cursor: pointer;
 -      line-height: inherit;
 -      position: absolute;
 -      background: none;
 -    }
 -  }
 -  
 -  .bx--accordion--start .bx--accordion__heading {
 -    background-image: linear-gradient(to bottom, #fafafa 0%, #ededed 100%);
 -  }
 -  
 -  .bx--accordion__content {
 -    margin: 0px;
 -    padding: 0px;
 -    flex: 1 1 auto;
 -    overflow-y: hidden;
 -    flex-direction: column;
 -    min-height: 0px; 
 -    .bx--inline-notification {
 -      margin-top: 0px;
 -      padding-top: 0px; 
 -      margin-bottom: 0px;
 -      min-height: 100px;
 -      height: auto;
 -      .bx--btn--ghost:hover, .bx--btn--ghost:active {
 -        background-color: inherit;
 -      }
 -      .bx--inline-notification__text-wrapper {
 -        flex-wrap: unset;
 -        .notification-time {
 -          left: 55px;
 -          position: absolute;
 -          bottom: 3px;
 -        }
 -        .bx--inline-notification__title {
 -          margin-bottom: 20px;
 -        }
 -        .subtitle-div {
 -          height: 100%;
 -          .subtitle-div-notification-menu {
 -            height: 80%;
 -          }
 -          .subtitle-div-notification-time {
 -            height: 20%;
 -          }
 -        }
 -      }
 -    }
 -  
 -    .read {
 -      .bx--inline-notification__title {
 -        font-weight: 400;
 -      }
 -      .bx--inline-notification__subtitle {
 -        font-weight: 200;
 -      }
 -    }
 -  }
 -     
 -}
 -
 -
 -.notification-module {
 -  display: inline;
 -}
 -
 -.notification-title {
 -  margin-left: 100px;
 -}    
 -.notification-content {
 -  display: flex;
 -  flex-direction: column;
 -  overflow-y: hidden;
 -  bottom: 0;
 -  margin-bottom: 0;
 -  position: absolute;
 -  top: 47px;
 -  width: 100%;
 -} 
 -
 -.events-count {
 -  display: flex;
 -  flex-direction: column;
 -}
 -  
 -.maxnotifications-text {
 -  margin-left: 60px;
 -  margin-top: 15px;
 -  display: flex;
 -  flex-direction: column;
 -  button {
 -    background: none!important;
 -    border: none;
 -    padding: 0!important;
 -    font-family: arial, sans-serif;
 -    text-decoration: none;
 -    cursor: pointer;
 -    color: #0f62fe;
 -  }
 -}
 -  
 -@media (min-width: 767px) {
 -  .notification-drawer.notification-drawer-expanded {
 -    left: 480px;
 -    width: inherit;
 -    .bx--inline-notification {
 -      max-width: inherit;
 -    }
 -    .notification-footer {
 -      width: 100%;
 -      padding-left: 450px;
 -    }
 -    .notification-time {
 -      padding-top: 0px !important;
 -    } 
 -
 -  }
 -}
 -
 -.notification-accordion-content {
 -  overflow-y: auto;
 -  height: 100%;
 -  .notification-content-messages {
 -    position: absolute;
 -    overflow-y: scroll;
 -    height: calc(100% - 115px);
 -  }
 -  .notification-footer {
 -    bottom: -65px;
 -    position: absolute;
 -    width: 450px;
 -    height: 115px;
 -    border-top: 1px solid #d1d1d1;
 -    Button {
 -      width: 225px;
 -      padding: calc(0.875rem - 3px) 35px;
 +    display: inline-block;
 +    height: 20px;
 +    width: 20px;
 +    text-align: center;
 +    i {
 +      align-self: center;
 +      color: #fff;
      }
    }
++<<<<<<< HEAD
++=======
+ }
+ 
+ .refresh-notifications {
+   display: flex;
+   .last-refresh-notification {
+     margin-top: 0px;
+     margin-left: 10px;
+   }
+   .date-notification {
+     margin-left: 20px;
+   }
+ }
++>>>>>>> faac0462d5... Merge pull request #7719 from kavyanekkalapu/add-status-provider-dashboard
```

kavyanekkalapu commented:
@Fryguy There are two conflicts. For one, i added `morphy/yes?` to dependent pr https://github.com/ManageIQ/manageiq-ui-classic/pull/7701. Another one, i cannot add dependent pr. There are too may prs and if i miss anything , it will cause lots of issues.  Is there an alternative to it other than redoing all these changes in morphy branch?
```
++=======
+ import PodsTrendChart from '../components/provider-dashboard-charts/pod-trend-charts';
+ import RecentHostGraph from '../components/provider-dashboard-charts/recent-host-chart';
+ import RecentVmGraph from '../components/provider-dashboard-charts/recent-vm-chart';
+ import RefreshDataNotification from '../components/refresh-data-notification';
++>>>>>>> faac0462d5... Merge pull request #7719 from kavyanekkalapu/add-status-provider-dashboard

Fryguy commented:
Which PR is the other dependent PR?

Fryguy commented:
Backported #7701

kavyanekkalapu commented:
> Which PR is the other dependent PR?

This is the dependent pr in which i converted provider dashboard charts https://github.com/ManageIQ/manageiq-ui-classic/pull/7658/files. recently i fixed some issues that are present in carbon-charts. if we merge this pr, i am not sure if that bugs affect Morphy. if those affects morphy, then i have to backport at least 4 more prs.

Fryguy commented:
Backported to `morphy`

```
commit fe519ef38c5e93f938a05e86324403b499f0e0fa (HEAD -> morphy, upstream/morphy)
Author: Adam Grare <adam@grare.com>
Date:   Fri May 7 15:28:58 2021 -0400

    Merge pull request #7719 from kavyanekkalapu/add-status-provider-dashboard

    Add last refresh status for provider dashboards

    (cherry picked from commit faac0462d5b018b95d8f0a56cdbe6ab4f5b365ea)
```
